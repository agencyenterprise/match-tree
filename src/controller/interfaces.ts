export const SeedsTypesAsArr = ['corn', 'green', 'black'];
export type SeedType = typeof SeedsTypesAsArr[number];

export interface ISeed {
  type: SeedType;
  row: number;
  col: number;
}

export interface IToClear {
  seeds: ISeed[];
  isBoost?: 'axe' | 'fork' | 'shovel';
}

export interface IMove {
  row: number;
  col: number;
  targetRow: number;
  targetCol: number;
  isValid: boolean;
}

export interface IGameState {
  seeds: (ISeed | null)[][];
  moves: IMove[];
  boardSize: number;
  isMoving: boolean;
}

export interface IGameFunctions {
  hasMove: () => { move: IMove; matching?: IToClear }[];
  tryMove(move: Omit<IMove, 'isValid'>): { move: IMove; matching?: IToClear };
  spawnSeeds: (args: {
    seeds?: (ISeed | null)[][] | null;
    allowMatching?: boolean;
  }) => ISeed[][];
  getLives: () => number;
  getScore: () => number;
  updateMatching: (matching: IToClear) => void;
  getEmptyCells: () => { col: number; rows: number[] }[];
}
