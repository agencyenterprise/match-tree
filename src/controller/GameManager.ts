import {
  IGameState,
  IGameFunctions,
  IMove,
  ISeed,
  IToClear,
  SeedsTypesAsArr,
  SeedType
} from './interfaces';
import * as _ from 'lodash';

export class GameManager implements IGameState, IGameFunctions {
  seeds: ISeed[][];
  moves: IMove[];
  boardSize: number;
  isMoving: boolean;
  minMatch: number;

  constructor({
    boardSize = 10,
    minMatch = 3,
    seeds
  }: {
    boardSize?: number;
    minMatch?: number;
    seeds?: ISeed[][];
  } = {}) {
    if (
      seeds &&
      (seeds.some((col) => col.length !== seeds.length) || !seeds.length)
    ) {
      throw new Error('seeds must be a square with some length');
    }

    this.moves = [];
    this.isMoving = false;
    this.minMatch = minMatch;
    if (!seeds) {
      this.boardSize = boardSize;
      do {
        this.seeds = this.spawnSeeds({ allowMatching: false });
      } while (!this.hasMove().length);
    } else {
      this.seeds = seeds;
      this.boardSize = seeds.length;
    }
  }

  getRandomSeed = () => {
    return SeedsTypesAsArr[Math.floor(Math.random() * SeedsTypesAsArr.length)];
  };

  hasMove = () => {
    const moves: { move: IMove; matching?: IToClear }[] = [];
    for (let i = 0; i < this.boardSize; i++) {
      for (let j = 0; j < this.boardSize; j++) {
        if (i + 1 < this.boardSize) {
          const move = this.tryMove({
            col: i,
            row: j,
            targetCol: i + 1,
            targetRow: j
          });
          if (move.move.isValid) {
            moves.push(move);
          }
        }

        if (j + 1 < this.boardSize) {
          const move = this.tryMove({
            col: i,
            row: j,
            targetCol: i,
            targetRow: j + 1
          });

          if (move.move.isValid) {
            moves.push(move);
          }
        }
      }
    }

    return _.uniq(moves);
  };
  tryMove = (
    move: Omit<IMove, 'isValid'>
  ): { move: IMove; matching?: IToClear } => {
    const colMov = Math.abs(move.col - move.targetCol);
    const rowMov = Math.abs(move.row - move.targetRow);
    if (
      (!colMov && !rowMov) || // Move to same spot
      (colMov && rowMov) || // Trying to move diagonally
      colMov > 1 || // Jumping a spot
      rowMov > 1
    ) {
      return { move: { ...move, isValid: false } };
    }

    const newSeeds = _.cloneDeep(this.seeds);
    const currentSeed = newSeeds[move.col][move.row];
    const targetSeed = newSeeds[move.targetCol][move.targetRow];
    newSeeds[move.col][move.row] = {
      ...targetSeed,
      col: move.col,
      row: move.row
    };
    newSeeds[move.targetCol][move.targetRow] = {
      ...currentSeed,
      col: move.targetCol,
      row: move.targetRow
    };

    const seeds = this.getMatching(newSeeds);
    if (seeds.length) {
      return { move: { ...move, isValid: true }, matching: { seeds } };
    } else {
      return { move: { ...move, isValid: false } };
    }
  };

  createEmptySeeds = (): null[][] => {
    const newSeeds = [];
    for (let i = 0; i < this.boardSize; ++i) {
      newSeeds.push(Array(this.boardSize).fill(null) as null[]);
    }
    return newSeeds;
  };

  getMatchingCols = (seeds: (ISeed | null)[][] = this.seeds) => {
    const seedsTransposed: (ISeed | null)[][] = this.createEmptySeeds();
    for (let i = 0; i < this.boardSize; ++i) {
      for (let j = 0; j < this.boardSize; ++j) {
        seedsTransposed[j][i] = seeds[i][j];
      }
    }
    return this.getMatchingRows(seedsTransposed);
  };
  getMatchingRows = (seeds: (ISeed | null)[][] = this.seeds) => {
    const matchingSeeds: ISeed[] = [];
    for (let i = 0; i < this.boardSize; ++i) {
      let latestSeedType: string | null = null;
      let counter = 1;
      const col = seeds[i];
      for (let j = 0; j < this.boardSize; ++j) {
        const seed = seeds[i][j];
        if (seed?.type === latestSeedType) {
          counter += 1;
          if (j >= this.boardSize - 1 && counter >= this.minMatch) {
            matchingSeeds.push(
              ...(col.slice(j - counter + 1, j + 1) as ISeed[])
            );
          }
        } else {
          if (counter >= this.minMatch) {
            matchingSeeds.push(...(col.slice(j - this.minMatch, j) as ISeed[]));
          }
          counter = seed ? 1 : 0;
        }
        latestSeedType = seed?.type ?? null;
      }
    }
    return matchingSeeds;
  };

  getMatching = (seeds: (ISeed | null)[][] = this.seeds) => {
    const matching = [
      ...this.getMatchingCols(seeds),
      ...this.getMatchingRows(seeds)
    ];
    return _.uniq(matching);
  };

  spawnSeeds = ({
    seeds = null,
    allowMatching = true
  }: {
    seeds?: (ISeed | null)[][] | null;
    allowMatching?: boolean;
  } = {}): ISeed[][] => {
    let newSeeds = seeds ? _.cloneDeep(seeds) : this.createEmptySeeds();
    for (let i = 0; i < this.boardSize; ++i) {
      for (let j = 0; j < this.boardSize; ++j) {
        const seed = seeds?.[i][j];
        if (seed == null) {
          do {
            newSeeds[i][j] = {
              col: i,
              row: j,
              type: this.getRandomSeed()
            };
          } while (!allowMatching && this.getMatching(newSeeds).length);
        } else {
          newSeeds[i][j] = seed;
        }
      }
    }
    return newSeeds as ISeed[][];
  };
  getLives = () => 0;
  getScore = () => 0;
  updateMatching!: (matching: IToClear) => 0;
  getEmptyCells!: () => { col: number; rows: number[] }[];
}
