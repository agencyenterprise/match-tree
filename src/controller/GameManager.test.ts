import { GameManager } from './GameManager';
import * as _ from 'lodash';
import { SeedType, ISeed } from './interfaces';

const shorthandTypeToSeedType = (type: 'b' | 'c' | 'g'): SeedType => {
  switch (type) {
    case 'b':
      return 'black';
    case 'c':
      return 'corn';
    case 'g':
      return 'green';
  }
};
const generateSeedsNullable = (...types: ('b' | 'c' | 'g' | null)[][]) =>
  types.map((col, i) =>
    col.map((type, j) =>
      type
        ? {
            col: i,
            row: j,
            type: shorthandTypeToSeedType(type)
          }
        : null
    )
  );
const generateSeeds = (...types: ('b' | 'c' | 'g')[][]) =>
  generateSeedsNullable(...types) as ISeed[][];

describe('on initialization', () => {
  it('create a board and inits a board', () => {
    const gamemanager = new GameManager({ boardSize: 10 });
    expect(gamemanager.boardSize).toEqual(10);
    expect(gamemanager.seeds.length).toEqual(10);
    expect(gamemanager.seeds[9].length).toEqual(10);
  });

  it('can override init seeds', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(['b', 'b', 'b'], ['c', 'g', 'b'], ['b', 'g', 'c'])
    });
    expect(gamemanager.boardSize).toEqual(3);
    expect(gamemanager.seeds.length).toEqual(3);
    expect(gamemanager.seeds[2].length).toEqual(3);
  });

  it('has cols and rows arranged correctly', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(['c', 'c', 'c'], ['c', 'g', 'b'], ['c', 'g', 'c'])
    });
    expect(gamemanager.seeds[1][2].type).toEqual('black');
  });

  it('validates seeds shape', () => {
    expect(
      () =>
        new GameManager({
          seeds: generateSeeds(['c', 'c'], ['c', 'g', 'b'], ['c', 'g', 'c'])
        })
    ).toThrow();
  });
});

describe('when checking if there is any move', () => {
  it('can find valid moves', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'b', 'g'],
        ['c', 'b', 'b', 'g'],
        ['g', 'c', 'g', 'c'],
        ['g', 'b', 'g', 'c']
      )
    });

    const moves = gamemanager.hasMove();
    expect(moves).toMatchSnapshot();
  });

  it('handles when there are no moves', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(['b', 'g', 'c'], ['c', 'b', 'c'], ['c', 'g', 'b'])
    });

    const moves = gamemanager.hasMove();
    expect(moves.length).toEqual(0);
  });
});

describe('when matching cols', () => {
  it('can get matching cols', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'g', 'g'],
        ['b', 'b', 'g', 'c'],
        ['b', 'g', 'g', 'g'],
        ['g', 'g', 'g', 'c']
      )
    });
    expect(gamemanager.getMatchingCols()).toMatchSnapshot();
  });

  it('can get matching cols when some are null', () => {
    const gamemanager = new GameManager({ boardSize: 4 });
    expect(
      gamemanager.getMatchingCols(
        generateSeedsNullable(
          ['b', 'c', 'g', 'g'],
          ['b', null, 'g', null],
          ['b', 'g', 'g', 'g'],
          ['g', 'g', 'g', 'c']
        )
      )
    ).toMatchSnapshot();
  });

  it('returns empty array if no matches', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'g', 'g'],
        ['c', 'b', 'g', 'c'],
        ['b', 'g', 'b', 'g'],
        ['g', 'c', 'g', 'c']
      )
    });
    expect(gamemanager.getMatchingCols()).toEqual([]);
  });
});

describe('when matching rows', () => {
  it('can get matching rows', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'b', 'b', 'g'],
        ['c', 'b', 'g', 'g'],
        ['g', 'g', 'g', 'g'],
        ['g', 'c', 'g', 'c']
      )
    });
    expect(gamemanager.getMatchingRows()).toMatchSnapshot();
  });

  it('can get matching rows when some are null', () => {
    const gamemanager = new GameManager({ boardSize: 4 });
    expect(
      gamemanager.getMatchingRows(
        generateSeedsNullable(
          ['b', 'b', 'b', 'g'],
          [null, 'b', null, 'g'],
          ['g', 'g', 'g', 'g'],
          ['g', 'c', 'g', 'c']
        )
      )
    ).toMatchSnapshot();
  });

  it('returns empty array if no matches', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'g', 'g'],
        ['c', 'b', 'g', 'c'],
        ['b', 'g', 'b', 'g'],
        ['g', 'c', 'g', 'c']
      )
    });
    expect(gamemanager.getMatchingRows()).toEqual([]);
  });
});

describe('when getting matching for both rows and cols', () => {
  it('can get valid matchings', () => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'b', 'g'],
        ['c', 'c', 'g', 'g'],
        ['g', 'c', 'g', 'c'],
        ['g', 'c', 'g', 'c']
      )
    });
    expect(gamemanager.getMatching()).toMatchSnapshot();
  });

  it('handles some seeds being null', () => {
    const gamemanager = new GameManager({ boardSize: 4 });
    expect(
      gamemanager.getMatching(
        generateSeedsNullable(
          ['b', 'c', 'b', 'g'],
          [null, 'c', null, 'g'],
          ['g', 'c', 'g', 'c'],
          ['g', 'c', 'g', 'c']
        )
      )
    ).toMatchSnapshot();
  });
});

it.each`
  col  | row  | targetCol | targetRow | isValid
  ${0} | ${0} | ${1}      | ${0}      | ${true}
  ${0} | ${2} | ${0}      | ${3}      | ${false}
  ${0} | ${0} | ${0}      | ${0}      | ${false}
  ${0} | ${0} | ${1}      | ${1}      | ${false}
  ${0} | ${0} | ${2}      | ${0}      | ${false}
  ${0} | ${0} | ${0}      | ${2}      | ${false}
`(
  'when trying move from ($col, $row) to ($targetCol, $targetRow), valid should be $isValid',
  ({ col, row, targetCol, targetRow, isValid }) => {
    const gamemanager = new GameManager({
      seeds: generateSeeds(
        ['b', 'c', 'b', 'g'],
        ['c', 'b', 'b', 'g'],
        ['g', 'c', 'g', 'c'],
        ['g', 'b', 'g', 'c']
      )
    });
    const result = gamemanager.tryMove({ col, row, targetCol, targetRow });
    expect(result.move.isValid).toEqual(isValid);
    expect(result).toMatchSnapshot();
  }
);

describe('when spawning new seeds', () => {
  it('can spawn new seeds to fill missing cells', () => {
    const gamemanager = new GameManager({
      boardSize: 4,
      minMatch: 3
    });
    const originalSeeds = generateSeedsNullable(
      ['b', 'c', null, 'g'],
      ['c', 'b', 'b', 'g'],
      [null, null, 'g', 'c'],
      ['g', 'b', null, null]
    );
    const newSeeds = gamemanager.spawnSeeds({ seeds: originalSeeds });
    expect(
      newSeeds.every((col, i) =>
        col.every(
          (seed, j) =>
            seed &&
            (originalSeeds[i][j] === null ||
              _.isEqual(originalSeeds[i][j], seed))
        )
      )
    ).toBeTruthy();
  });

  it('can spawn new seeds when matching is not allowed', () => {
    const gamemanager = new GameManager({
      boardSize: 4,
      minMatch: 3
    });
    const originalSeeds = generateSeedsNullable(
      ['b', 'c', null, 'g'],
      ['c', 'b', 'b', 'g'],
      [null, null, 'g', 'c'],
      ['g', 'b', null, null]
    );
    const newSeeds = gamemanager.spawnSeeds({
      seeds: originalSeeds,
      allowMatching: false
    });
    expect(gamemanager.getMatching(newSeeds).length).toEqual(0);
    expect(
      newSeeds.every((col, i) =>
        col.every(
          (seed, j) =>
            seed &&
            (originalSeeds[i][j] === null ||
              _.isEqual(originalSeeds[i][j], seed))
        )
      )
    ).toBeTruthy();
  });

  it('can spawn new seeds on board init', () => {
    const gamemanager = new GameManager();
    expect(gamemanager.getMatching().length).toEqual(0);
    expect(gamemanager.seeds.length).toEqual(10);
    gamemanager.seeds.every((col) => expect(col.length).toEqual(10));
    expect(
      gamemanager.seeds.every((col) => col.every((seed) => seed))
    ).toBeTruthy();
    expect(gamemanager.hasMove().length).toBeTruthy();
  });
});
