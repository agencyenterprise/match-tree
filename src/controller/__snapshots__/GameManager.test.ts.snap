// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when checking if there is any move can find valid moves 1`] = `
Array [
  Object {
    "matching": Object {
      "seeds": Array [
        Object {
          "col": 1,
          "row": 0,
          "type": "black",
        },
        Object {
          "col": 1,
          "row": 1,
          "type": "black",
        },
        Object {
          "col": 1,
          "row": 2,
          "type": "black",
        },
      ],
    },
    "move": Object {
      "col": 0,
      "isValid": true,
      "row": 0,
      "targetCol": 1,
      "targetRow": 0,
    },
  },
  Object {
    "matching": Object {
      "seeds": Array [
        Object {
          "col": 0,
          "row": 0,
          "type": "black",
        },
        Object {
          "col": 0,
          "row": 1,
          "type": "black",
        },
        Object {
          "col": 0,
          "row": 2,
          "type": "black",
        },
      ],
    },
    "move": Object {
      "col": 0,
      "isValid": true,
      "row": 1,
      "targetCol": 1,
      "targetRow": 1,
    },
  },
  Object {
    "matching": Object {
      "seeds": Array [
        Object {
          "col": 0,
          "row": 1,
          "type": "corn",
        },
        Object {
          "col": 1,
          "row": 1,
          "type": "corn",
        },
        Object {
          "col": 2,
          "row": 1,
          "type": "corn",
        },
      ],
    },
    "move": Object {
      "col": 1,
      "isValid": true,
      "row": 0,
      "targetCol": 1,
      "targetRow": 1,
    },
  },
  Object {
    "matching": Object {
      "seeds": Array [
        Object {
          "col": 1,
          "row": 2,
          "type": "green",
        },
        Object {
          "col": 2,
          "row": 2,
          "type": "green",
        },
        Object {
          "col": 3,
          "row": 2,
          "type": "green",
        },
      ],
    },
    "move": Object {
      "col": 1,
      "isValid": true,
      "row": 2,
      "targetCol": 1,
      "targetRow": 3,
    },
  },
  Object {
    "matching": Object {
      "seeds": Array [
        Object {
          "col": 0,
          "row": 3,
          "type": "green",
        },
        Object {
          "col": 1,
          "row": 3,
          "type": "green",
        },
        Object {
          "col": 2,
          "row": 3,
          "type": "green",
        },
      ],
    },
    "move": Object {
      "col": 2,
      "isValid": true,
      "row": 2,
      "targetCol": 2,
      "targetRow": 3,
    },
  },
]
`;

exports[`when getting matching for both rows and cols can get valid matchings 1`] = `
Array [
  Object {
    "col": 0,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 1,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 2,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 3,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 1,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 3,
    "row": 2,
    "type": "green",
  },
]
`;

exports[`when getting matching for both rows and cols handles some seeds being null 1`] = `
Array [
  Object {
    "col": 0,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 1,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 2,
    "row": 1,
    "type": "corn",
  },
  Object {
    "col": 3,
    "row": 1,
    "type": "corn",
  },
]
`;

exports[`when matching cols can get matching cols 1`] = `
Array [
  Object {
    "col": 0,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 1,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 2,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 1,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 3,
    "row": 2,
    "type": "green",
  },
]
`;

exports[`when matching cols can get matching cols when some are null 1`] = `
Array [
  Object {
    "col": 0,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 1,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 2,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 1,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 3,
    "row": 2,
    "type": "green",
  },
]
`;

exports[`when matching rows can get matching rows 1`] = `
Array [
  Object {
    "col": 0,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 1,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 2,
    "type": "black",
  },
  Object {
    "col": 2,
    "row": 0,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 1,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 3,
    "type": "green",
  },
]
`;

exports[`when matching rows can get matching rows when some are null 1`] = `
Array [
  Object {
    "col": 0,
    "row": 0,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 1,
    "type": "black",
  },
  Object {
    "col": 0,
    "row": 2,
    "type": "black",
  },
  Object {
    "col": 2,
    "row": 0,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 1,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 2,
    "type": "green",
  },
  Object {
    "col": 2,
    "row": 3,
    "type": "green",
  },
]
`;

exports[`when trying move from (0, 0) to (0, 0), valid should be false 1`] = `
Object {
  "move": Object {
    "col": 0,
    "isValid": false,
    "row": 0,
    "targetCol": 0,
    "targetRow": 0,
  },
}
`;

exports[`when trying move from (0, 0) to (0, 2), valid should be false 1`] = `
Object {
  "move": Object {
    "col": 0,
    "isValid": false,
    "row": 0,
    "targetCol": 0,
    "targetRow": 2,
  },
}
`;

exports[`when trying move from (0, 0) to (1, 0), valid should be true 1`] = `
Object {
  "matching": Object {
    "seeds": Array [
      Object {
        "col": 1,
        "row": 0,
        "type": "black",
      },
      Object {
        "col": 1,
        "row": 1,
        "type": "black",
      },
      Object {
        "col": 1,
        "row": 2,
        "type": "black",
      },
    ],
  },
  "move": Object {
    "col": 0,
    "isValid": true,
    "row": 0,
    "targetCol": 1,
    "targetRow": 0,
  },
}
`;

exports[`when trying move from (0, 0) to (1, 1), valid should be false 1`] = `
Object {
  "move": Object {
    "col": 0,
    "isValid": false,
    "row": 0,
    "targetCol": 1,
    "targetRow": 1,
  },
}
`;

exports[`when trying move from (0, 0) to (2, 0), valid should be false 1`] = `
Object {
  "move": Object {
    "col": 0,
    "isValid": false,
    "row": 0,
    "targetCol": 2,
    "targetRow": 0,
  },
}
`;

exports[`when trying move from (0, 2) to (0, 3), valid should be false 1`] = `
Object {
  "move": Object {
    "col": 0,
    "isValid": false,
    "row": 2,
    "targetCol": 0,
    "targetRow": 3,
  },
}
`;
